pants-plugins/examples/bash/shunit2_test_runner.py

```diff
@@ -0,0 +1,181 @@
+from pants.engine.target import Dependencies, Sources, TransitiveTargets
+from pants.engine.unions import UnionRule
+from pants.util.logging import LogLevel
+
+from examples.bash.bash_setup import BashProgram, BashSetup
+from examples.bash.target_types import BashSources, BashTestSources
+
+
+@dataclass(frozen=True)
+class Shunit2FieldSet(TestFieldSet):
+    required_fields = (BashTestSources,)
+
+    sources: BashTestSources
+    dependencies: Dependencies
+
+
+@dataclass(frozen=True)
+class TestSetupRequest:
+    field_set: Shunit2FieldSet
+
+
+@dataclass(frozen=True)
+class TestSetup:
+    process: Process
+
+
+@rule(level=LogLevel.DEBUG)
+async def setup_shunit2_for_target(
+    request: TestSetupRequest, bash_program: BashProgram, bash_setup: BashSetup
+) -> TestSetup:
+    # Because shunit2 is a simple Bash file, we download it using `DownloadFile`. Normally, we
+    # would install the test runner through `ExternalTool`. See
+    # https://www.pantsbuild.org/v2.0/docs/rules-api-installing-tools and
+    # https://www.pantsbuild.org/v2.0/docs/rules-api-file-system.
+    shunit2_script_request = Get(
+        Digest,
+        DownloadFile(
+            url="https://raw.githubusercontent.com/kward/shunit2/b9102bb763cc603b3115ed30a5648bf950548097/shunit2",
+            expected_digest=Digest(
+                "1f11477b7948150d1ca50cdd41d89be4ed2acd137e26d2e0fe23966d0e272cc5",
+                40987,
+            ),
+        ),
+    )
+
+    transitive_targets_request = Get(
+        TransitiveTargets, Addresses([request.field_set.address])
+    )
+
+    shunit2_script, transitive_targets = await MultiGet(
+        shunit2_script_request, transitive_targets_request
+    )
+
+    # We need to include all relevant transitive dependencies in the environment. We also get the
+    # test's sources so that we can check that it has `source ./shunit2` at the bottom of it.
+    #
+    # Because we might modify the test files, we leave the tests out of
+    # `dependencies_source_files_request` by using `transitive_targets.dependencies` instead of
+    # `transitive_targets.closure`. This makes sure that we don't accidentally include the
+    # unmodified test files and the modified test files in the same input. See
+    # https://www.pantsbuild.org/v2.0/docs/rules-api-and-target-api.
+    dependencies_source_files_request = Get(
+        SourceFiles,
+        SourceFilesRequest(
+            (tgt.get(Sources) for tgt in transitive_targets.dependencies),
+            for_sources_types=(BashSources, FilesSources, ResourcesSources),
+        ),
+    )
+    test_source_files_request = Get(
+        SourceFiles, SourceFilesRequest([request.field_set.sources])
+    )
+    dependencies_source_files, test_source_files = await MultiGet(
+        dependencies_source_files_request, test_source_files_request
+    )
+
+    # To check if the test files already have `source ./shunit2` in them, we need to look at the
```

Comment:
TBH I'm not sure this is necessary, and it complicates the example. We're supposed to be "a thin wrapper around tools", and "we will edit your source files behind your back" is very not-that.  

I think if someone is writing a shunit2 test, and must already be aware of how shunit2 works, then expecting them to add necessary boilerplate is totally fine, rather then being too-clever-by-half.

Reply:
Hm, would you expect us to still download shunit2 automatically?

Would the user need to put `source ./shunit2` (sibling file), or we should attempt to understand `source ../../shunit2`?

Reply:
Ah, I see, that does muddy the waters. 

What are users expected to do in general, check shunit2 into their repos?  That would, for example, guarantee a specific version.

Reply:
Yeah, I think so. And then reference it relative to where it's checked in.

But this is a huge selling point of Pants: we will get the tools for you, whether it be shunit2 or Pytest. That's really nice imo that now you can write a test anywhere in your repo without having to remember where you checked in the `./shunit2` script.

Reply:
Fair. And we don't prevent people from doing the manual thing if they want.

Reply:
Exactly